src/search/pdbs/cegar.cc

```diff
@@ -0,0 +1,589 @@
+        }
+    }
+    return flaws;
+}
+
+FlawList CEGAR::get_flaws_for_projection(
+    int collection_index, const State &concrete_init) {
+    const Projection &projection = *projection_collection[collection_index];
+    if (projection.is_unsolvable()) {
+        utils::g_log << "task is unsolvable." << endl;
+        utils::exit_with(utils::ExitCode::SEARCH_UNSOLVABLE);
+    }
+
+    State current(concrete_init);
+    current.unpack();
+    FlawList flaws = apply_plan(collection_index, current);
+    if (flaws.empty()) {
+        if (task_properties::is_goal_state(task_proxy, current)) {
+            if (verbosity >= utils::Verbosity::VERBOSE) {
+                utils::g_log << "plan led to a concrete goal state: ";
+            }
+            if (blacklisted_variables.empty()) {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "there are no blacklisted variables, "
+                        "task solved." << endl;
+                }
+                concrete_solution_index = collection_index;
+            } else {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "there are blacklisted variables, "
+                        "marking projection as solved." << endl;
+                }
+                projection_collection[collection_index]->mark_as_solved();
+            }
+        } else {
+            if (verbosity >= utils::Verbosity::VERBOSE) {
+                utils::g_log << "plan did not lead to a goal state: ";
+            }
+            for (const FactPair &goal : goals) {
+                int goal_var_id = goal.var;
+                if (current[goal_var_id].get_pair() != goal && !blacklisted_variables.count(goal_var_id)) {
+                    flaws.emplace_back(collection_index, goal_var_id);
+                }
+            }
+            if (flaws.empty()) {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "there are no non-blacklisted goal variables "
+                        "left, marking projection as solved." << endl;
+                }
+                projection_collection[collection_index]->mark_as_solved();
+            } else {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "raising goal violation flaw(s)." << endl;
+                }
+            }
+        }
+    }
+    return flaws;
+}
+
+FlawList CEGAR::get_flaws() {
+    FlawList flaws;
+    State concrete_init = task_proxy.get_initial_state();
+    for (size_t collection_index = 0;
+         collection_index < projection_collection.size(); ++collection_index) {
+        if (projection_collection[collection_index] &&
+            !projection_collection[collection_index]->is_solved()) {
+            FlawList new_flaws =
+                get_flaws_for_projection(collection_index, concrete_init);
+            if (concrete_solution_index != -1) {
```

Comment:
We need to pass the index to `get_flaws_for_projection` because that in turn passes it to `apply_plan`. Note that indices are part of what a flaw constitutes: a variable plus the index of the projection in which it is a flaw. So we cannot simplify the loop unfortunately. I also dislike pairs as return values. 

If anything, I could imagine dropping `new_flaws` and instead passing `flaws` to `get_flaws_for_projection` which then modifies it and returns a single boolean. We could even go a step further and also pass `flaws` as an argument to `get_flaws`, which could then return a boolean if the problem is solved. This would allow us removing the `concrete_solution_index`. Instead, we could exit the main method `compute_pattern_collection` if the returned value is true. Otherwise, assert that flaws is non-empty and proceed. 

I know that passing non-const empty parameters for modifying them isn't that beautiful either, but I think it is a more common pattern and better than lists (or pairs) of return values. What do you think?

Reply:
I went ahead and implemented the suggestion. Please let me know if you don't like it.